api/src/clojure/org/akvo/flow_api/unilog/unilog.clj

```diff
@@ -142,26 +143,37 @@
                                       ["SELECT id AS offset FROM event_log WHERE id = ?" offset]))]
         (boolean (:offset result)))))
 
-(defn process-unilog-events [offset db instance-id remote-api]
+(defn read-event-log
+  [db offset]
+  (jdbc/reducible-query db
+                        ["SELECT id, payload::text AS payload FROM event_log WHERE id > ? ORDER BY id ASC LIMIT 300" offset]
+                        {:auto-commit? false :fetch-size 300}))
+
+(defn process-unilog-events [offset db instance-id remote-api email]
   (ds/with-remote-api remote-api instance-id
     (let [ds (DatastoreServiceFactory/getDatastoreService)
-          events (process-new-events
-                   (jdbc/reducible-query db
-                                         ["SELECT id, payload::text AS payload FROM event_log WHERE id > ? ORDER BY id ASC LIMIT 300" offset]
-                                         {:auto-commit? false :fetch-size 300}))
+          events (process-new-events (read-event-log db offset))
+          user-id (user/id-by-email-or-throw-error remote-api instance-id email)
+          authorized-forms (set (map ds/id (su/list-forms user-id)))
+          authorized-forms-to-load (set/intersection authorized-forms (:forms-to-load events))
           form-id->form (reduce (fn [acc form-id]
                                   (assoc acc form-id (su/get-form-definition (long form-id)
                                                                              {:include-survey-id? true})))
                                 {}
-                                (:forms-to-load events))
-          events-2 (after-forms-loaded events form-id->form) ;; TODO: get form definition from cache
+                                authorized-forms-to-load)
+          authorized-surveys (set (map ds/id (su/list* user-id)))
+          authorized-events (filter-events-by-authorization events form-id->form authorized-forms authorized-surveys)
           form-instances (doall
                            (mapcat
                              (fn [to-load]
                                (form-instance/by-ids ds (:form to-load) (:form-instance-ids to-load)))
-                             (:form-instances-to-load events-2)))
-          data-point-changed (doall (data-point/by-ids ds (:data-point-changed events-2)))]
-      (assoc events-2 :form-instance-changed form-instances :data-point-changed data-point-changed))))
+                             (:form-instances-to-load authorized-events)))
+          data-point-changed (doall (data-point/by-ids ds (map :id (:data-point-changed authorized-events))))
```

Comment:
The result from `filter-events-by-authorization` should be manipulated as little as possible, because `filter-events-by-authorization` is properly unit tested, while this code is not. Hence I would suggest that `filter-events-by-authorization` should just return the `id` if that is the only thing needed.